Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscSignatureValidator.kt

```diff
@@ -2,16 +2,144 @@ package de.rki.coronawarnapp.covidcertificate.signature.core
+import java.security.cert.CertificateFactory
+import java.security.cert.CertificateNotYetValidException
+import java.security.cert.X509Certificate
 import javax.inject.Inject
 
 @Reusable
-class DscSignatureValidator @Inject constructor() {
+class DscSignatureValidator @Inject constructor(
+    private val dscRepository: DscRepository
+) {
 
-    suspend fun isSignatureValid(dscData: DscData, certificateData: DccData<*>): Boolean {
-        Timber.tag(TAG).d("isSignatureValid(certificateData=%s)", certificateData.header)
-        // TODO
-        return true
+    init {
+        Security.addProvider(BouncyCastleProvider()) // For SHA256withRSA/PSS
+    }
+
+    private val vcOids = setOf(
+        "1.3.6.1.4.1.1847.2021.1.2",
+        "1.3.6.1.4.1.0.1847.2021.1.2"
+    )
+    private val tcOids = setOf(
+        "1.3.6.1.4.1.1847.2021.1.1",
+        "1.3.6.1.4.1.0.1847.2021.1.1"
+    )
+    private val rcOids = setOf(
+        "1.3.6.1.4.1.1847.2021.1.3",
+        "1.3.6.1.4.1.0.1847.2021.1.3"
+    )
+
+    private val oidSet = vcOids + tcOids + rcOids
+
+    /**
+     * @throws InvalidHealthCertificateException if validation fail, otherwise it is OK!
+     */
+    suspend fun validateSignature(dccData: DccData<*>) {
+        val dscData = dscRepository.dscData.first()
+        Timber.tag(TAG).d("validateSignature(dscListSize=%s)", dscData.dscList.size)
+
+        findDscCertificate(dscData, dccData.dscMessage).apply {
+            validate()
+            checkOidsIntersect(dccData)
+        }
+    }
+
+    private fun findDscCertificate(
+        dscData: DscData,
+        dscMessage: DscMessage
+    ): X509Certificate {
+        val toVerify = dscMessage.signedPayload()
+        val filteredDscSet = dscData.dscList.filter { it.kid == dscMessage.kid }
+        Timber.d("filteredDscSetSize=${filteredDscSet.size}")
+
+        val matchedDscSet = when {
+            filteredDscSet.isEmpty() || dscMessage.kid.isEmpty() -> dscData.dscList
+            else -> filteredDscSet
+        }
+        Timber.d("matchedDscSetSize=${matchedDscSet.size}")
+
+        var x509Certificate: X509Certificate? = null
+        for (dsc in matchedDscSet) {
+            val dscCertificate = dsc.toX509certificate()
+            val (publicKey, signature) = when (dscMessage.algorithm) {
+                ES256 -> dscCertificate.publicKey to dscMessage.signature.toByteArray().toECDSAVerifier()
+                PS256 -> dscCertificate.publicKey.toRsaPublicKey() to dscMessage.signature.toByteArray()
+            }
+
+            try {
+                val valid = Signature.getInstance(dscMessage.algorithm.algName).verify(publicKey, toVerify, signature)
+                Timber.d("Dsc certificate (${dsc.kid}) is valid=$valid")
+
+                if (valid) {
+                    x509Certificate = dscCertificate
+                    break
+                }
+            } catch (ignored: Exception) {
+                // Ignore errors / continue
```

Comment:
Could we maybe add a small  `.v` log message here? 
This seems like a critical part that could fails for various reasons that we would be happy about to see in a debug log.

Reply:
Please check the specs , it states on no logs are required

Reply:
@mlenkeit Why should nothing be logged here?

Reply:
signature verification can naturally fail, so the idea was not to pollute the log in that case, especially when there's a larger set of DSCs to check. But in reality, the 99% case is 1 DSC and pass, so proper logging for the 1% makes sense. See spec update.